builtin/game/hud.lua

```diff
@@ -0,0 +1,235 @@
+]]--
+local registered_elements = {}
+local update_events = {}
+local function register_builtin_hud_element(name, def)
+	registered_elements[name] = def
+	if def.event then
+		local events = update_events[def.event]
+		if events then
+			table.insert(events, name)
+		else
+			update_events[def.event] = {name}
+		end
+	end
+end
+
+-- Stores HUD ids for all players
+local hud_ids = {}
+
+-- Updates one element
+-- In case the element is already added, it does only call the hud_change function from
+-- registered_elements. (To completely update the element call player:hud_remove(id) first.)
+local function update_element(player, player_hud_ids, elem_name, flags)
+	local def = registered_elements[elem_name]
+	local id = player_hud_ids[elem_name]
+
+	if def.show_elem and not def.show_elem(player, flags, id) then
+		if id then
+			player:hud_remove(id)
+		end
+		return
+	end
+
+	if not id then
+		id = player:hud_add(def.elem_def)
+		player_hud_ids[elem_name] = id
+	end
+
+	if def.hud_change then
+		def.hud_change(player, id)
+	end
+end
+
+-- Updates all elements
+local function update_hud(player)
+	local flags = player:hud_get_flags()
+	local playername = player:get_player_name()
+	local player_hud_ids = hud_ids[playername]
+	if not player_hud_ids then
+		player_hud_ids = {}
+		hud_ids[playername] = player_hud_ids
+	end
+	for elem_name, _ in pairs(registered_elements) do
+		update_element(player, player_hud_ids, elem_name, flags)
+	end
+end
+
+local function player_event_handler(player, eventname)
+	assert(player:is_player())
+
+	if eventname == "hud_changed" or eventname == "properties_changed" then
+		update_hud(player)
+		return true
+	end
+
+	-- Custom events
+	local to_update = update_events[eventname]
+	if to_update then
+		local flags = player:hud_get_flags()
+		local playername = player:get_player_name()
+		local player_hud_ids = hud_ids[playername]
+		if not player_hud_ids then
+			player_hud_ids = {}
+			hud_ids[playername] = player_hud_ids
+		end
+		for _, elem_name in ipairs(to_update) do
+			update_element(player, player_hud_ids, elem_name, flags)
+		end
+		return true
+	end
+
+	return false
+end
+
+-- Returns true if successful, otherwise false,
+-- but currently the return value is not part of the lua API.
+function core.hud_replace_builtin(elem_name, definition)
+	assert(type(definition) == "table")
+
+	local registered = registered_elements[elem_name]
+	if not registered then
+		return false
+	end
+
+	registered.elem_def = table.copy(definition)
+
+	for playername, player_hud_ids in pairs(hud_ids) do
+		local player = core.get_player_by_name(playername)
+		local id = player_hud_ids[elem_name]
+		if player and id then
+			player:hud_remove(id)
+			update_element(player, player_hud_ids, elem_name, player:hud_get_flags())
+		end
+	end
+	return true
+end
+
+local function cleanup_builtin_hud(player)
+	local name = player:get_player_name()
+	if name == "" then
+		return
+	end
```

Comment:
How, possibly, would a player with the name `""` leave?

Reply:
I took this from the old version, I guess it happens when the player somehow does not exist anymore.
Would you prefer a `minetest.is_player(player)` check?

Reply:
This function is only called as a callback given to `register_on_leaveplayer`. If Minetest deleted or invalidated objects before calling `register_on_leaveplayer` callbacks, we'd have a serious problem and many mods would be broken. What I'd prefer is removing the check completely, as it gives the false impression of being necessary.